src/main.rs

```diff
@@ -47,6 +52,68 @@ async fn main() {
     // Determine port to listen on
     let port = get_port(&config);
 
+    // Find or generate secret keys
+    let public_keys_path = config.get_str("crypto.keys.publicpath").unwrap();
+    let secret_keys_path = config.get_str("crypto.keys.privatepath").unwrap();
+    let filter = fs_io::FsFilterer::new();
+    let public_keys_filtered = filter.dir(&public_keys_path, 1, 1, Some(&"pub")).unwrap();
+    for e in public_keys_filtered.io_errors.iter() {
+        println!("{}", e)
+    }
+    let private_keys_filtered = filter.dir(&secret_keys_path, 1, 1, Some(&"key")).unwrap();
+    for e in private_keys_filtered.io_errors.iter() {
+        println!("{}", e)
+    }
+
+    let mut default_key_name = match config.get_str("crypto.keys.defaultkeyname") {
+        Ok(dkn) => {
+            if dkn == "" {
+                "admin".to_owned()
+            } else {
+                dkn
+            }
+        }
+        Err(e) => {
+            println!(
+                "error getting crypto.keys.defaultkeyname, defaulting to 'admin': {}",
+                e
+            );
+            "admin".to_owned()
+        }
+    };
+
+    let (pk, sk) = match public_keys_filtered
+        .paths
+        .iter()
+        .find(|&path| path.file_stem().unwrap() == "admin")
+    {
+        Some(path) => {
+            let mut pk_file = File::open(path).unwrap();
+            let mut sk_file = File::open(
+                PathBuf::from(&secret_keys_path)
+                    .join(path.file_stem().unwrap().to_str().unwrap().to_owned() + ".key"),
+            )
+            .unwrap();
+            let mut pk_arr: [u8; 64] = [0; 64];
+            let mut sk_arr: [u8; 64] = [0; 64];
+            assert!(pk_file.read(&mut pk_arr).unwrap() == 64);
+            assert!(sk_file.read(&mut sk_arr).unwrap() == 64);
+            (pk_arr, sk_arr)
+        }
+        None => {
+            let keys = SodiumOxideKeypairGenerator::create().unwrap();
+            let pk = keys.0;
+            let sk = keys.1;
+            let new_pk_path = PathBuf::from(&public_keys_path).join(PathBuf::from("admin.pub"));
```

Comment:
same as before, should it be `default_key_name` + ".pub" here?

Reply:
yes